---
title: "qcGenes Draft"
author: "JF Fontaine"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    theme: united
    toc: true
    toc_depth: 2
    toc_float:
      collapsed: false
      smooth_scroll: true
    number_sections: true
    fig_width: 7
    fig_height: 6
    fig_caption: yes
  pdf_document:
    toc: true
    number_sections: true
    extra_dependencies:
    - grffile
---

<!-- TO DO -->
<!-- Impact factor VS Bias -->


```{r include=FALSE}
library(tidyverse)
library(yaml)
library(knitr)
options(readr.num_columns = 0)
```

```{r include=FALSE}
config.path                   <- file.path("..", "..", "config", "config.yaml")
config.data                   <- yaml.load_file(config.path)

datasets.stats.path         <- file.path("..", "..", "output", "main", "qualityCorGenes", "datasets.stats.tsv")
datasets.stats              <- read_tsv(datasets.stats.path)
datasets.total.n            <- dim(datasets.stats)[1]
datasets.selected.n         <- sum(datasets.stats$selected)
datasets.selected.quality.n <- sum(datasets.stats$p_group_cor<config.data$MAX_DATASET_P_VS_GROUP_CORRELATION)
datasets.selected.deg.n     <- sum(datasets.stats$n_degs>config.data$MIN_DATASET_N_DEG)

datasets.stats <- datasets.stats %>% filter(n_degs>50 & n_samples>10)
datasets.selected.filtered.n <- nrow(datasets.stats)

datasets.dunn.pos <- sum((datasets.stats$dunn_noout-datasets.stats$dunn_all)>0 & datasets.stats$selected==TRUE)
datasets.dunn2.pos <- sum((datasets.stats$dunn2_noout-datasets.stats$dunn2_all)>0 & datasets.stats$selected==TRUE)
datasets.gamma.pos <- sum((datasets.stats$pgamma_noout-datasets.stats$pgamma_all)>0 & datasets.stats$selected==TRUE)

datasets.dunn.pos.nocor <- sum((datasets.stats$dunn_noout_nocor-datasets.stats$dunn_noout_norand)>0 & datasets.stats$selected==TRUE)
datasets.dunn2.pos.nocor <- sum((datasets.stats$dunn2_noout_nocor-datasets.stats$dunn2_noout_norand)>0 & datasets.stats$selected==TRUE)
datasets.gamma.pos.nocor <- sum((datasets.stats$pgamma_noout_nocor-datasets.stats$pgamma_noout_norand)>0 & datasets.stats$selected==TRUE)

median.p <- median((datasets.stats %>% filter(selected==TRUE))$p_group_cor)

selected.lowP.datasets  <- datasets.stats %>% filter(p_group_cor<=median.p & selected==TRUE)
selected.highP.datasets <- datasets.stats %>% filter(p_group_cor>median.p & selected==TRUE)
selected.lowP.datasets.n <- nrow(selected.lowP.datasets)
selected.highP.datasets.n <- nrow(selected.highP.datasets)

datasets.lowP.dunn.pos <- sum((selected.lowP.datasets$dunn_noout-selected.lowP.datasets$dunn_all)>0)
datasets.lowP.dunn2.pos <- sum((selected.lowP.datasets$dunn2_noout-selected.lowP.datasets$dunn2_all)>0)
datasets.lowP.gamma.pos <- sum((selected.lowP.datasets$pgamma_noout-selected.lowP.datasets$pgamma_all)>0)
datasets.highP.dunn.pos <- sum((selected.highP.datasets$dunn_noout-selected.highP.datasets$dunn_all)>0)
datasets.highP.dunn2.pos <- sum((selected.highP.datasets$dunn2_noout-selected.highP.datasets$dunn2_all)>0)
datasets.highP.gamma.pos <- sum((selected.highP.datasets$pgamma_noout-selected.highP.datasets$pgamma_all)>0)

datasets.lowP.dunn.pos.nocor <- sum((selected.lowP.datasets$dunn_noout_nocor-selected.lowP.datasets$dunn_noout_norand)>0)
datasets.lowP.dunn2.pos.nocor <- sum((selected.lowP.datasets$dunn2_noout_nocor-selected.lowP.datasets$dunn2_noout_norand)>0)
datasets.lowP.gamma.pos.nocor <- sum((selected.lowP.datasets$pgamma_noout_nocor-selected.lowP.datasets$pgamma_noout_norand)>0)

datasets.highP.dunn.pos.nocor <- sum((selected.highP.datasets$dunn_noout_nocor-selected.highP.datasets$dunn_noout_norand)>0)
datasets.highP.dunn2.pos.nocor <- sum((selected.highP.datasets$dunn2_noout_nocor-selected.highP.datasets$dunn2_noout_norand)>0)
datasets.highP.gamma.pos.nocor <- sum((selected.highP.datasets$pgamma_noout_nocor-selected.highP.datasets$pgamma_noout_norand)>0)

#selected.lowP.datasets.n
#round(100*(datasets.lowP.dunn.pos.nocor/selected.lowP.datasets.n), 0)
#round(100*(datasets.lowP.dunn2.pos.nocor/selected.lowP.datasets.n), 0)
#round(100*(datasets.lowP.gamma.pos.nocor/selected.lowP.datasets.n), 0)
#selected.highP.datasets.n
#round(100*(datasets.highP.dunn.pos.nocor/selected.highP.datasets.n), 0)
#round(100*(datasets.highP.dunn2.pos.nocor/selected.highP.datasets.n), 0)
#round(100*(datasets.highP.gamma.pos.nocor/selected.highP.datasets.n), 0)


```


<!-- ======================================================================= -->
# Introduction
<!-- ======================================================================= -->

<!-- ----------------------------------------------------------------------- -->
## Objective
<!-- ----------------------------------------------------------------------- -->

* Evaluate impact of low quality samples in the analysis of NGS datasets
* Use many datasets to identify recurrent patterns, such as:
  * signal (e.g. genes or peaks) biased by quality in several independant datasets
  * boost in samples clustering after removal of lowest quality samples

<!-- ----------------------------------------------------------------------- -->
## Methods
<!-- ----------------------------------------------------------------------- -->

**`r datasets.total.n` datasets** from GEO (GSExxxx) were selected with the following criteria:

* assay: single-ended or pair-ended RNA-seq
* samples: human cells or tissues
* design: control vs disease, unpaired samples (?)
* availability: public data from GEO and SRA databases

Processing tools:

* Low-quality probabilities of NGS samples (P<sub>low</sub>): SeqQscorer (incl. fastqc, bowtie2, chippeakanno, and chipseeker)
* Differential genes: RASflow pipeline (salmon, DESeq2) 
* Disease genes annotation: GS2D tool (literature db mining)


<!-- ======================================================================= -->
# Results
<!-- ======================================================================= -->

<!-- ----------------------------------------------------------------------- -->
## Datasets quality
<!-- ----------------------------------------------------------------------- -->

**`r datasets.selected.n` (`r round(100*(datasets.selected.n)/datasets.total.n, 0)`%) datasets were selected** for further analysis based on:

* Minimal number of significant differential genes = `r config.data$MIN_DATASET_N_DEG`
* Maxmimal correlation coefficient of quality vs sample groups = `r config.data$MAX_DATASET_P_VS_GROUP_CORRELATION`

**`r datasets.total.n-datasets.selected.n` datasets were excluded**:

* `r datasets.total.n-datasets.selected.deg.n` with low number of signigicant differential genes: impossible to use for testing clustering boost
* `r datasets.total.n-datasets.selected.quality.n` with high association of sample groups with P<sub>low</sub>: impossible to differentiate disease-relevant and quality-biased genes

**Notably, `r round(100*(datasets.total.n-datasets.selected.quality.n)/datasets.total.n, 2)`% (n=`r datasets.total.n-datasets.selected.quality.n`) of the datasets in the literature are highly biased by quality** and unlikely to be debiased by correction methods (Soneson et al. PLoS One 2014)


![*color: dataset selection; bar annotation: # significant differential genes*](../../output/main/qualityCorGenes/cor.group.pdf)

<!-- ----------------------------------------------------------------------- -->
## Quality and differential genes
<!-- ----------------------------------------------------------------------- -->

In the analysis of a given dataset, the number of differential genes is expected 
to depend on the number of samples. The higher the number of samples, the higher 
the power of the statistical tests. 
But, to which extend would the samples quality impact this number?

* **The expected inflation of the number of differential genes in relation to the number of samples is drastically increased in high-bias datasets**


![](../../output/main/qualityCorGenes/deg.fdrOnly.samples.pdf)

* **For large effects (fold change>2), the bias still impacts the number of differential genes in high-bias datasets, although differential genes are less/not expected to increase with the number of samples**

![](../../output/main/qualityCorGenes/deg.samples.pdf)

* **Number of differential genes does not depend on bias in moderate-sized datasets**

![](../../output/main/qualityCorGenes/deg.plow.pdf)

<!-- ----------------------------------------------------------------------- -->
## Quality-filtering and clustering across selected datasets
<!-- ----------------------------------------------------------------------- -->

**P<sub>low</sub>** is used to identify low-quality samples in datasets, i.e. 
outliers, or quality-biased genes (gene expresssion correlates with sample's quality). 

The impact of having either outlier samples or quality-biased genes in a dataset is evaluated by comparing the clustering of the samples by group (control or treatment/disease) on a PCA in 2 different analyses as follows:

* outlier samples removal: compare samples clustering before and after removal
* biased genes removal:  compare samples clustering after outlier removal and either removal of quality biased genes or same number of random genes

* **Dunn index and Pearson/Normalised Gamma index** are used to evaluate clustering of samples by groups on a PCA. 
* **For both type of index: the higher the index, the better the clustering**.
* **Positive difference denotes better clustering after removal of outlier samples or quality-biased genes**
* **Recurrence in multiple datasets would demonstrate a notable impact of quality on gene-expression**


<!-- ### Removal of outlier samples -->
<!-- ----------------------------------------------------------------------- -->

<!-- Datasets with positive difference out of `r datasets.selected.filtered.n` selected datasets: -->

<!-- * Dunn index:  `r datasets.dunn.pos`  (`r round(100*(datasets.dunn.pos/datasets.selected.n), 0)`%) -->
<!-- * Dunn2 index:  `r datasets.dunn2.pos`  (`r round(100*(datasets.dunn2.pos/datasets.selected.n), 0)`%) -->
<!-- * Gamma index:  `r datasets.gamma.pos`  (`r round(100*(datasets.gamma.pos/datasets.selected.n), 0)`%) -->

<!-- If we use the **median P<sub>low</sub> value of `r round(median.p,3)`** to divide the `r datasets.selected.n` selected datasets in 2 categories:  -->

<!-- **Low-P<sub>low</sub> datasets** with positive difference out of `r selected.lowP.datasets.n` selected datasets: -->

<!-- * Dunn index:  `r datasets.lowP.dunn.pos`  (`r round(100*(datasets.lowP.dunn.pos/selected.lowP.datasets.n), 0)`%) -->
<!-- * Dunn2 index:  `r datasets.lowP.dunn2.pos`  (`r round(100*(datasets.lowP.dunn2.pos/selected.lowP.datasets.n), 0)`%) -->
<!-- * Gamma index:  `r datasets.lowP.gamma.pos`  (`r round(100*(datasets.lowP.gamma.pos/selected.lowP.datasets.n), 0)`%) -->

<!-- **High-P<sub>low</sub> datasets** with positive difference out of `r selected.highP.datasets.n` selected datasets: -->

<!-- * Dunn index:  `r datasets.highP.dunn.pos`  (`r round(100*(datasets.highP.dunn.pos/selected.lowP.datasets.n), 0)`%) -->
<!-- * Dunn2 index:  `r datasets.highP.dunn2.pos`  (`r round(100*(datasets.highP.dunn2.pos/selected.lowP.datasets.n), 0)`%) -->
<!-- * Gamma index:  `r datasets.highP.gamma.pos`  (`r round(100*(datasets.highP.gamma.pos/selected.lowP.datasets.n), 0)`%) -->


<!-- **Based on the 3 clustering indices, removal of quality-biased genes improves samples clustering in average in `r round(100*mean(c(datasets.dunn.pos/datasets.selected.n, datasets.dunn2.pos/datasets.selected.n, datasets.gamma.pos/datasets.selected.n)),2)`% of the `r datasets.selected.n` selected datasets.** -->


<!-- ### Removal of quality-biased genes -->
<!-- ----------------------------------------------------------------------- -->

<!-- Datasets with positive difference out of `r datasets.selected.filtered.n` selected datasets: -->

<!-- * Dunn index:  `r datasets.dunn.pos.nocor`  (`r round(100*(datasets.dunn.pos.nocor/datasets.selected.n), 0)`%) -->
<!-- * Dunn2 index:  `r datasets.dunn2.pos.nocor`  (`r round(100*(datasets.dunn2.pos.nocor/datasets.selected.n), 0)`%) -->
<!-- * Gamma index:  `r datasets.gamma.pos.nocor`  (`r round(100*(datasets.gamma.pos.nocor/datasets.selected.n), 0)`%) -->

<!-- If we use the **median P<sub>low</sub> value of `r round(median.p,3)`** to divide the `r datasets.selected.n` selected datasets in 2 categories:  -->

<!-- **Low-P<sub>low</sub> datasets** with positive difference out of `r selected.lowP.datasets.n` selected datasets: -->

<!-- * Dunn index:  `r datasets.lowP.dunn.pos.nocor`  (`r round(100*(datasets.lowP.dunn.pos.nocor/selected.lowP.datasets.n), 0)`%) -->
<!-- * Dunn2 index:  `r datasets.lowP.dunn2.pos.nocor`  (`r round(100*(datasets.lowP.dunn2.pos.nocor/selected.lowP.datasets.n), 0)`%) -->
<!-- * Gamma index:  `r datasets.lowP.gamma.pos.nocor`  (`r round(100*(datasets.lowP.gamma.pos.nocor/selected.lowP.datasets.n), 0)`%) -->

<!-- **High-P<sub>low</sub> datasets** with positive difference out of `r selected.highP.datasets.n` selected datasets: -->

<!-- * Dunn index:  `r datasets.highP.dunn.pos.nocor`  (`r round(100*(datasets.highP.dunn.pos.nocor/selected.lowP.datasets.n), 0)`%) -->
<!-- * Dunn2 index:  `r datasets.highP.dunn2.pos.nocor`  (`r round(100*(datasets.highP.dunn2.pos.nocor/selected.lowP.datasets.n), 0)`%) -->
<!-- * Gamma index:  `r datasets.highP.gamma.pos.nocor`  (`r round(100*(datasets.highP.gamma.pos.nocor/selected.lowP.datasets.n), 0)`%) -->

<!-- **Based on the 3 clustering indices, removal of quality-biased genes improves samples clustering in average in `r round(100*mean(c(datasets.lowP.dunn.pos.nocor/selected.lowP.datasets.n, datasets.lowP.dunn2.pos.nocor/selected.lowP.datasets.n, datasets.lowP.gamma.pos.nocor/selected.lowP.datasets.n)),2)`% of the low-P<sub>low</sub> selected datasets.** -->


<!-- ![*diff: difference; dunn: Dunn index; pgamma: Normalised or Pearson's gamma index*](../../output/main/qualityCorGenes/pca.genes.removal.indices.pdf) -->


### Clustering overview
<!-- ----------------------------------------------------------------------- -->

Observing the distribution of positive and negative differences (Dunn1 and Gamma indices), we observe that:

* outlier removal is better or not detrimental for most of the datasets.
* biased genes removal: the lower the bias, the better

![](../../output/main/qualityCorGenes/pca.indices.plow.scatter.pdf)
![](../../output/main/qualityCorGenes/pca.indices.positives.pdf)

### Clustering details
<!-- ----------------------------------------------------------------------- -->

Observing the distribution of positive and negative differences, we observe that:


* outlier removal is better or not detrimental for most of the datasets.
    * the few detrimental examples are related to small- or medium-sized datasets
* biased genes removal: the lower the bias, the better
    * small impact on big datasets
    * most of the detrimental examples are related to small- or medium-sized datasets

![](../../output/main/qualityCorGenes/pca.indices.plow.scatter.extraFacets.pdf)

![](../../output/main/qualityCorGenes/pca.indices.positives.extraFacets.pdf)


<!-- ----------------------------------------------------------------------- -->
## Quality-biased genes recurrence
<!-- ----------------------------------------------------------------------- -->
    
Quality-biased genes have their expression correlating with low-quality 
probabilities across the samples in a dataset. 

* **Recurrence in multiple datasets would demonstrate a systemic bias**
* **Gene functions, enriched for those genes, may point to biological events**

### Distributions

* **Positive correlations: the higher the gene expression, the lower the quality**
* **Negative correlations: the higher the gene expression, the higher the quality**

#### Low-bias datasets

```{r echo=FALSE, out.width="50%"}
myimages <- c("../../output/main/qualityCorGenes/pos.cor.genes.pdf",
              "../../output/main/qualityCorGenes/pos.cor.all.genes.pdf")
include_graphics(myimages)
```

```{r echo=FALSE, out.width="50%"}
myimages <- c("../../output/main/qualityCorGenes/neg.cor.genes.pdf",
              "../../output/main/qualityCorGenes/neg.cor.all.genes.pdf")
include_graphics(myimages)
```

#### High-bias datasets

```{r echo=FALSE, out.width="50%"}
myimages <- c("../../output/main/qualityCorGenes/pos.cor.genes.highBias.pdf",
              "../../output/main/qualityCorGenes/pos.cor.all.genes.highBias.pdf")
include_graphics(myimages)
```

```{r echo=FALSE, out.width="50%"}
myimages <- c("../../output/main/qualityCorGenes/neg.cor.genes.highBias.pdf",
              "../../output/main/qualityCorGenes/neg.cor.all.genes.highBias.pdf")
include_graphics(myimages)
```

### Enrichment analysis

#### Positive correlations
**No-bias datasets**
![](../../output/main/qualityCorGenes/fgsea/fgsea.pos.cure.noBias.gsea.plot.pdf)

**Low-bias datasets**
![](../../output/main/qualityCorGenes/fgsea/fgsea.pos.cure.gsea.plot.pdf)
**High-bias datasets**
![](../../output/main/qualityCorGenes/fgsea/fgsea.pos.cure.highBias.gsea.plot.pdf)

#### Negative correlations
**No-bias datasets**
![](../../output/main/qualityCorGenes/fgsea/fgsea.neg.cure.noBias.gsea.plot.pdf)

**Low-bias datasets**
![](../../output/main/qualityCorGenes/fgsea/fgsea.neg.cure.gsea.plot.pdf)

**High-bias datasets**
![](../../output/main/qualityCorGenes/fgsea/fgsea.neg.cure.highBias.gsea.plot.pdf)


<!-- ----------------------------------------------------------------------- -->
# Outlook
<!-- ----------------------------------------------------------------------- -->

Investigate trimming, batch correction, and alternate dea method (e.g. EdgeR).


<!-- ----------------------------------------------------------------------- -->
# Supplementary material
<!-- ----------------------------------------------------------------------- -->

<!-- ----------------------------------------------------------------------- -->
## Low-quality probability (P<sub>low</sub>)
<!-- ----------------------------------------------------------------------- -->

Derived by machine learning on multiple quality features, Low-quality 
probabilities are expected to be complex as they do not correlate with a single 
feature (see seqQscorer manuscript). The higher the probability, the lower the 
expected quality.

* probabilities should not systematically correlate with data size 
* probabilities could catch batch-processing effect

<details>
  <summary><font color="#FF0000">Click for Answer</font></summary>

### P<sub>low</sub> and data size

**If not systematicly correlated with data size, P<sub>low</sub> would be confirmed to be complex.**

![*p: low-quality probabilities; bases: data size in bases (nucleotides); bases: data size in bytes; cor: correlation coefficient; ctrl: in control samples; treat: in treated/disease samples*](../../output/main/qualityCorGenes/cor.size.pdf)

<!--
AUTOMATIC FIGURE NUMBERING WORKS ONLY FOR PDF OUTPUT
Figure \ref{fig:fig1}
```{r fig1, fig.cap="\\label{fig:fig1}This is a caption"}
# All defaults
include_graphics("../../output/main/qualityCorGenes/cor.size.pdf")
```
-->

### P<sub>low</sub> and batch effect

Batch effect is known to be associated with differences in quality. 
**If P<sub>low</sub> is associated with batch effect, it would be a validation**.
It may be used to identify batches or to enable identification of within-batch outliers.

```{r echo=FALSE, out.width="50%", out.extra='style="background-color: #BBBBBB; padding:2px; display: inline-block;"'}
#myimages <- c("../../output/batched/qualityVsExp/GSE163214/GSE163214.batch.pdf",
#              "../../output/batched/qualityVsExp/GSE173078/GSE173078.batch.pdf",
#              "../../output/batched/qualityVsExp/GSE81871/GSE81871.batch.pdf",
#              "../../output/batched/qualityVsExp/GSE82177/GSE82177.batch.pdf",
#              "../../output/batched/qualityVsExp/GSE120099/GSE120099.batch.pdf")
#include_graphics(myimages)
```
</details> 

